refactor(filetype)!: allow vim.filetype.match to use different strategies

gpanders opened the pull request:
This enables vim.filetype.match to match based on a buffer (most accurate) or simply a filename or file contents, which are less accurate but may still be useful for some scenarios.

When matching based on a buffer, the buffer's name and contents are both used to do full filetype matching. When using a filename, if the file exists the file is loaded into a buffer and full filetype detection is performed. If the file does not exist then filetype matching is only performed against the filename itself. Content-based matching does the equivalent of scripts.vim, and matches solely based on file contents without any information from the name of the file itself (e.g. for shebangs).

clason commented:
Hmm, I'm torn on this. On the one hand, this is a welcome simplification; on the other hand, one of the goals of this feature is to allow getting(!) the filetype in other situations -- I know nvim-treesitter is interested in this for codeblock injections.

Of course, without a proper file, many of the more involved checks just won't work, no matter how this code is refactored. But `vim.filetype` should still be usable even in this restricted situation. Maybe we could add a "fast" option that skips all content checks (and, possibly, only looks at the extension)?

clason commented:
Would it make sense to do the same for the individual functions, i.e., pushing getting the `path` from the `bufnr` to only those branches where it's needed?

clason commented:
On the third hand, the file name is already available for free from the autocommand `args`, so it feels a bit wasteful to throw it away just to retrieve it again down the road...

gpanders commented:
Yes I've been flipping and flopping too, which is why I added the `needs-discussion` label.

One concern I have is that when the `name` and `bufnr` fields are decoupled, how should `match()` behave when the `name` and `bufnr` don't match? e.g. if buffer 1 has name `main.c`, what does `vim.filetype.match('hello.py', 1)` even mean? It's a nonsensical use of the API and there's no way to enforce consistency, so it falls to the user to use it properly which is heading into footgun territory.

muniter commented:
Since it's up for discussion here's a take.

Only take **one of the two parameters** using a table.

If it's a name `vim.filetype.match({file = '/tmp/afile'})`, use it as a path, resolve it and perform detection over it, one could load it as an unlisted buffer for that, or modify the wrapper get_lines function to read from the path.

If it's a buffer `vim.filetype.match({buffer = 5})` do what is currently done.

Now nvim provides a nice function for filetype detection of any given path or buffer.

clason commented:
Alternative (as this is a bit too verbose for my taste): make it a single positional argument `string|number` and dispatch on that; this would also remove the need for throwing an error if both buffer and path are provided?

clason commented:
Hmm, that's a good point, and a strong argument for making (only) the buffer number a mandatory argument. (In this case, I'd punt getting the path to the last possible moment: the `filetype.detect` functions.)

The question of "fast-tracking" a simple name to filetype lookup (where `nil` is acceptable if it can't be decided without looking at contents) would have to be addressed with a separate function then.

matveyt commented:
Maybe then just create two public "match" functions - one by bufnr, another by filename that share one common implementation? And if the implementation is private it is not much of a concern if it deduces something from argument type, or accepts extra option, etc.etc.

gpanders commented:
This is the interface I'm leaning toward:

A single `vim.filetype.match` function that accepts a table with one of the 3 keys: `bufnr`, `filename`, `contents`. Each key is mutually exclusive, it is an error to supply more than one.

`bufnr` does basically what this PR does: it matches against the file name of the buffer and for the functions that check buffer contents, runs those checks as well.

`filename` only matches against the file path, and does not require an underlying buffer. **Detection functions are not run**, this will *only* match against string filetypes that match against the given filename. This would need to be documented that using `filename` *may not provide a match* and that using `bufnr` is the best solution, if possible.

`contents` takes a string corresponding to the contents of a file (or maybe just the first `N` lines?) and matches strictly based on those contents. This is essentially what `scripts.lua` would use and is useful for e.g. shebang filetype detection.

This is relevant to #18710 as well: if this interface is adopted then the new `scripts.lua`  would simply call `local ft = vim.filetype.match({contents = getlines(bufnr, 1, -1)}); vim.bo[bufnr].filetype = ft`.

Thoughts?

clason commented:
Yes, I think this is a very good approach. 

As this is a Lua API, we might consider a little magic (as a treat), where not providing any option defaults to the current buffer? (But then again, we've been deprecating "nil means current buffer", so this could be counterproductive.)

clason commented on runtime/lua/vim/filetype.lua:
```diff
@@ -2214,29 +2215,85 @@ local function match_pattern(name, path, tail, pat)
+    error('Only one of "buf", "filename", or "contents" must be given')
+  end
+
   -- When fired from the main filetypedetect autocommand the {bufnr} argument is omitted, so we use
```
Given the validation above, this is no longer true?

clason commented on runtime/lua/vim/filetype.lua:
```diff
@@ -2214,29 +2215,85 @@ local function match_pattern(name, path, tail, pat)
 
-  name = normalize_path(name)
+  if not bufnr and vim.loop.fs_access(name, 'R') then
+    -- When matching against a filename and the file is readable, load it into a buffer so that we
```
That's not how I understood the documentation -- that seemed to imply that _only_ the filename would be used, ever?

(Which I think is the right approach.)

clason commented:
@Conni2461 @fdschmidt93 @theHamsta could you give this a look and comment on whether that API is useful for plenary/telescope and nvim-treesitter, respectively?

gpanders committed:
lint

gpanders committed:
refactor(filetype)!: allow vim.filetype.match to use different strategies

This enables vim.filetype.match to match based on a buffer (most
accurate) or simply a filename or file contents, which are less accurate
but may still be useful for some scenarios.

When matching based on a buffer, the buffer's name and contents are both
used to do full filetype matching. When using a filename, if the file
exists the file is loaded into a buffer and full filetype detection is
performed. If the file does not exist then filetype matching is only
performed against the filename itself. Content-based matching does the
equivalent of scripts.vim, and matches solely based on file contents
without any information from the name of the file itself (e.g. for
shebangs).

gpanders committed:
fixup doc comment example

smjonas commented on runtime/lua/vim/filetype.lua:
```diff
@@ -2214,29 +2221,74 @@ local function match_pattern(name, path, tail, pat)
+  if not (bufnr or name) then
+    -- Sanity check: this should not happen
+    assert(contents, 'contents should be non-nil when bufnr and filename are nil')
+    -- TODO: "scripts.lua" content matching
```
In `scripts.lua`, in some cases (like when the first line matches `^csh\>`) the `sh()` function in `detect.lua` is called to retrieve the correct filetype and do additional checks. In this case, won't `bufnr` and `path` be required then?

clason replied:
We talked about `pcall`ing the functions if no `bufnr` is provided so we can, in particular, correctly match `filename='eval.cpp'`. We could do the same here?

clason replied:
But I agree that this means characterizing `scripts.lua` as "content matching" is slightly misleading -- it is just the last (mostly content-based) step in the matching pipeline, independently of the argument (although it can -- and should -- indeed be skipped for `filename`?)
